File: goblinframework-core/src/main/java/org/goblinframework/core/serialization/hessian/Hessian2Serializer.java
package org.goblinframework.core.serialization.hessian;

import com.caucho.hessian.io.Hessian2Input;
import com.caucho.hessian.io.Hessian2Output;
import org.goblinframework.api.common.Singleton;
import org.goblinframework.core.exception.GoblinSerializationException;
import org.goblinframework.core.serialization.Serializer;
import org.goblinframework.core.serialization.SerializerMode;
import org.jetbrains.annotations.NotNull;

import java.io.*;

@Singleton
final public class Hessian2Serializer implements Serializer {

  public static final Hessian2Serializer INSTANCE = new Hessian2Serializer();

  private Hessian2Serializer() {
  }

  @NotNull
  @Override
  public SerializerMode mode() {
    return SerializerMode.HESSIAN2;
  }

  @Override
  public void serialize(@NotNull Object obj, @NotNull OutputStream outStream) {
    if (!(obj instanceof Serializable)) {
      throw GoblinSerializationException.requiredSerializable(obj);
    }
    Hessian2Output ho = null;
    try {
      ho = GoblinHessianFactory.INSTANCE.createHessian2Output(outStream);
      ho.writeObject(obj);
      ho.flush();
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    } finally {
      GoblinHessianFactory.INSTANCE.freeHessian2Output(ho);
    }
  }

  @NotNull
  @Override
  public byte[] serialize(@NotNull Object obj) {
    if (!(obj instanceof Serializable)) {
      throw GoblinSerializationException.requiredSerializable(obj);
    }
    try (ByteArrayOutputStream bos = new ByteArrayOutputStream(512)) {
      serialize(obj, bos);
      return bos.toByteArray();
    } catch (IOException ex) {
      throw new GoblinSerializationException(ex);
    }
  }

  @NotNull
  @Override
  public Object deserialize(@NotNull InputStream inStream) {
    Hessian2Input hi = null;
    try {
      hi = GoblinHessianFactory.INSTANCE.createHessian2Input(inStream);
      return hi.readObject();
    } catch (Exception ex) {
      throw new GoblinSerializationException(ex);
    } finally {
      GoblinHessianFactory.INSTANCE.freeHessian2Input(hi);
    }
  }

  @NotNull
  @Override
  public Object deserialize(@NotNull byte[] bs) {
    try (ByteArrayInputStream bis = new ByteArrayInputStream(bs)) {
      return deserialize(bis);
    } catch (IOException ex) {
      throw new GoblinSerializationException(ex);
    }
  }
}

File: goblinframework-serialization-hessian/src/main/java/org/goblinframework/serialization/hessian/provider/Hessian2Serializer.kt
package org.goblinframework.serialization.hessian.provider

import com.caucho.hessian.io.Hessian2Input
import com.caucho.hessian.io.Hessian2Output
import org.goblinframework.api.serialization.Serialization
import org.goblinframework.api.serialization.Serializer0
import java.io.InputStream
import java.io.OutputStream

class Hessian2Serializer : Serialization {

  override fun getSerializer(): Serializer0 {
    return Serializer0.HESSIAN2
  }

  override fun serialize(obj: Any, outStream: OutputStream) {
    var ho: Hessian2Output? = null
    try {
      ho = GoblinHessianFactory.INSTANCE.createHessian2Output(outStream)
      ho.writeObject(obj)
      ho.flush()
    } finally {
      GoblinHessianFactory.INSTANCE.freeHessian2Output(ho)
    }
  }

  override fun deserialize(inStream: InputStream): Any {
    var hi: Hessian2Input? = null
    try {
      hi = GoblinHessianFactory.INSTANCE.createHessian2Input(inStream)
      return hi.readObject()
    } finally {
      GoblinHessianFactory.INSTANCE.freeHessian2Input(hi)
    }
  }
}